File: src/marktoflow/core/scheduler.py
"""
Scheduler for marktoflow framework.

Handles cron-based scheduling of workflow execution.
"""

from __future__ import annotations

import asyncio
import logging
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Any, Callable

from marktoflow.core.models import Workflow
from marktoflow.core.parser import WorkflowParser

logger = logging.getLogger(__name__)


@dataclass
class ScheduledJob:
    """A scheduled workflow job."""

    id: str
    workflow_path: str
    schedule: str  # Cron expression
    timezone: str = "UTC"
    enabled: bool = True
    last_run: datetime | None = None
    next_run: datetime | None = None
    run_count: int = 0
    inputs: dict[str, Any] = field(default_factory=dict)


class CronParser:
    """
    Simple cron expression parser.

    Supports: minute hour day_of_month month day_of_week
    Special values: * (any), */N (every N), N-M (range), N,M (list)
    """

    @staticmethod
    def parse(expression: str) -> dict[str, list[int]]:
        """
        Parse a cron expression into component values.

        Args:
            expression: Cron expression (5 fields)

        Returns:
            Dict with minute, hour, day, month, weekday lists
        """
        parts = expression.strip().split()

        if len(parts) != 5:
            raise ValueError(f"Invalid cron expression: {expression}")

        ranges = {
            "minute": (0, 59),
            "hour": (0, 23),
            "day": (1, 31),
            "month": (1, 12),
            "weekday": (0, 6),
        }

        field_names = ["minute", "hour", "day", "month", "weekday"]
        result = {}

        for i, (name, (min_val, max_val)) in enumerate(zip(field_names, ranges.values())):
            result[name] = CronParser._parse_field(parts[i], min_val, max_val)

        return result

    @staticmethod
    def _parse_field(field: str, min_val: int, max_val: int) -> list[int]:
        """Parse a single cron field."""
        values = set()

        for part in field.split(","):
            if part == "*":
                values.update(range(min_val, max_val + 1))
            elif "/" in part:
                base, step = part.split("/")
                step = int(step)
                if base == "*":
                    start = min_val
                else:
                    start = int(base)
                values.update(range(start, max_val + 1, step))
            elif "-" in part:
                start, end = map(int, part.split("-"))
                values.update(range(start, end + 1))
            else:
                values.add(int(part))

        return sorted(values)

    @staticmethod
    def matches(expression: str, dt: datetime) -> bool:
        """Check if a datetime matches a cron expression."""
        try:
            parsed = CronParser.parse(expression)
        except ValueError:
            return False

        return (
            dt.minute in parsed["minute"]
            and dt.hour in parsed["hour"]
            and dt.day in parsed["day"]
            and dt.month in parsed["month"]
            and dt.weekday() in parsed["weekday"]
        )

    @staticmethod
    def next_run(expression: str, after: datetime | None = None) -> datetime | None:
        """
        Calculate the next run time for a cron expression.

        Args:
            expression: Cron expression
            after: Start time (defaults to now)

        Returns:
            Next matching datetime or None if invalid
        """
        from datetime import timedelta

        if after is None:
            after = datetime.now()

        try:
            parsed = CronParser.parse(expression)
        except ValueError:
            return None

        # Start from next minute
        current = after.replace(second=0, microsecond=0) + timedelta(minutes=1)

        # Search up to 1 year ahead
        max_iterations = 366 * 24 * 60

        for _ in range(max_iterations):
            if CronParser.matches(expression, current):
                return current
            current = current + timedelta(minutes=1)

        return None


class Scheduler:
    """
    Workflow scheduler.

    Manages scheduled execution of workflows based on cron expressions.
    """

    def __init__(
        self,
        workflows_path: str | Path = ".marktoflow/workflows",
        config: dict[str, Any] | None = None,
    ) -> None:
        """
        Initialize the scheduler.

        Args:
            workflows_path: Path to workflows directory
            config: Scheduler configuration
        """
        self.workflows_path = Path(workflows_path)
        self.config = config or {}
        self._jobs: dict[str, ScheduledJob] = {}
        self._running = False
        self._parser = WorkflowParser()
        self._callbacks: list[Callable[[ScheduledJob, Any], None]] = []

    def load_schedules(self) -> int:
        """
        Load scheduled workflows from the workflows directory.

        Returns:
            Number of scheduled jobs loaded
        """
        if not self.workflows_path.exists():
            return 0

        count = 0
        for workflow_file in self.workflows_path.glob("*.md"):
            try:
                workflow = self._parser.parse_file(workflow_file)
                schedules = self._extract_schedules(workflow)

                for i, schedule in enumerate(schedules):
                    job_id = f"{workflow.metadata.id}_{i}"
                    job = ScheduledJob(
                        id=job_id,
                        workflow_path=str(workflow_file),
                        schedule=schedule["schedule"],
                        timezone=schedule.get("timezone", "UTC"),
                        enabled=schedule.get("enabled", True),
                        next_run=CronParser.next_run(schedule["schedule"]),
                    )
                    self._jobs[job_id] = job
                    count += 1

            except Exception as e:
                logger.warning(f"Failed to load workflow {workflow_file}: {e}")

        return count

    def _extract_schedules(self, workflow: Workflow) -> list[dict[str, Any]]:
        """Extract schedule triggers from a workflow."""
        schedules = []

        for trigger in workflow.triggers:
            if trigger.type == "schedule" and trigger.enabled:
                schedule_config = trigger.config
                if "schedule" in schedule_config:
                    schedules.append(schedule_config)

        return schedules

    def add_job(self, job: ScheduledJob) -> None:
        """Add a scheduled job."""
        self._jobs[job.id] = job

    def remove_job(self, job_id: str) -> bool:
        """Remove a scheduled job."""
        if job_id in self._jobs:
            del self._jobs[job_id]
            return True
        return False

    def get_job(self, job_id: str) -> ScheduledJob | None:
        """Get a scheduled job by ID."""
        return self._jobs.get(job_id)

    def list_jobs(self) -> list[ScheduledJob]:
        """List all scheduled jobs."""
        return list(self._jobs.values())

    def on_job_complete(
        self,
        callback: Callable[[ScheduledJob, Any], None],
    ) -> None:
        """Register a callback for job completion."""
        self._callbacks.append(callback)

    async def start(self) -> None:
        """Start the scheduler loop."""
        self._running = True
        logger.info(f"Scheduler started with {len(self._jobs)} jobs")

        while self._running:
            now = datetime.now()

            for job in self._jobs.values():
                if not job.enabled:
                    continue

                if job.next_run and now >= job.next_run:
                    # Time to run this job
                    await self._execute_job(job)

                    # Update next run time
                    job.last_run = now
                    job.run_count += 1
                    job.next_run = CronParser.next_run(job.schedule, now)

            # Sleep until next minute
            await asyncio.sleep(60 - now.second)

    async def _execute_job(self, job: ScheduledJob) -> None:
        """Execute a scheduled job."""
        logger.info(f"Executing scheduled job: {job.id}")

        try:
            # Load workflow
            workflow = self._parser.parse_file(job.workflow_path)

            # Create engine and execute
            from marktoflow.core.engine import WorkflowEngine

            engine = WorkflowEngine(config=self.config)
            result = await engine.execute(workflow, inputs=job.inputs)

            # Notify callbacks
            for callback in self._callbacks:
                try:
                    callback(job, result)
                except Exception as e:
                    logger.error(f"Callback error: {e}")

            if result.success:
                logger.info(f"Job {job.id} completed successfully")
            else:
                logger.error(f"Job {job.id} failed: {result.error}")

        except Exception as e:
            logger.error(f"Failed to execute job {job.id}: {e}")

    def stop(self) -> None:
        """Stop the scheduler."""
        self._running = False
        logger.info("Scheduler stopped")

    async def run_once(self) -> dict[str, Any]:
        """
        Check and run any due jobs once (non-blocking).

        Returns:
            Dict with execution results
        """
        results = {}
        now = datetime.now()

        for job in self._jobs.values():
            if not job.enabled:
                continue

            if job.next_run and now >= job.next_run:
                await self._execute_job(job)
                job.last_run = now
                job.run_count += 1
                job.next_run = CronParser.next_run(job.schedule, now)
                results[job.id] = {"status": "executed", "time": now}

        return results

File: packages/core/src/scheduler.ts
/**
 * Scheduler for marktoflow v2.0
 *
 * Handles cron-based scheduling of workflow execution.
 */

// ============================================================================
// Types
// ============================================================================

export interface ScheduledJob {
  id: string;
  workflowPath: string;
  schedule: string; // Cron expression
  timezone: string;
  enabled: boolean;
  lastRun: Date | null;
  nextRun: Date | null;
  runCount: number;
  inputs: Record<string, unknown>;
}

export interface CronFields {
  minute: number[];
  hour: number[];
  day: number[];
  month: number[];
  weekday: number[];
}

export type JobCallback = (job: ScheduledJob) => Promise<void>;

// ============================================================================
// Cron Parser
// ============================================================================

export class CronParser {
  /**
   * Parse a cron expression into component values.
   * Format: minute hour day month weekday
   * Special values: * (any), * /N (every N), N-M (range), N,M (list)
   */
  static parse(expression: string): CronFields {
    const parts = expression.trim().split(/\s+/);

    if (parts.length !== 5) {
      throw new Error(`Invalid cron expression: ${expression}. Expected 5 fields.`);
    }

    const ranges: Record<keyof CronFields, [number, number]> = {
      minute: [0, 59],
      hour: [0, 23],
      day: [1, 31],
      month: [1, 12],
      weekday: [0, 6],
    };

    const fieldNames: (keyof CronFields)[] = ['minute', 'hour', 'day', 'month', 'weekday'];
    const result: CronFields = {
      minute: [],
      hour: [],
      day: [],
      month: [],
      weekday: [],
    };

    for (let i = 0; i < fieldNames.length; i++) {
      const name = fieldNames[i];
      const [min, max] = ranges[name];
      result[name] = this.parseField(parts[i], min, max);
    }

    return result;
  }

  private static parseField(field: string, minVal: number, maxVal: number): number[] {
    const values = new Set<number>();

    for (const part of field.split(',')) {
      if (part === '*') {
        for (let i = minVal; i <= maxVal; i++) {
          values.add(i);
        }
      } else if (part.includes('/')) {
        const [base, stepStr] = part.split('/');
        const step = parseInt(stepStr, 10);
        const start = base === '*' ? minVal : parseInt(base, 10);

        for (let i = start; i <= maxVal; i += step) {
          values.add(i);
        }
      } else if (part.includes('-')) {
        const [startStr, endStr] = part.split('-');
        const start = parseInt(startStr, 10);
        const end = parseInt(endStr, 10);

        for (let i = start; i <= end; i++) {
          values.add(i);
        }
      } else {
        values.add(parseInt(part, 10));
      }
    }

    return Array.from(values).sort((a, b) => a - b);
  }

  /**
   * Check if a date matches a cron expression.
   */
  static matches(expression: string, date: Date): boolean {
    try {
      const fields = this.parse(expression);

      return (
        fields.minute.includes(date.getMinutes()) &&
        fields.hour.includes(date.getHours()) &&
        fields.day.includes(date.getDate()) &&
        fields.month.includes(date.getMonth() + 1) &&
        fields.weekday.includes(date.getDay())
      );
    } catch {
      return false;
    }
  }

  /**
   * Calculate the next run time for a cron expression.
   */
  static nextRun(expression: string, after?: Date): Date | null {
    const start = after || new Date();

    try {
      this.parse(expression);
    } catch {
      return null;
    }

    // Start from next minute
    const current = new Date(start);
    current.setSeconds(0, 0);
    current.setMinutes(current.getMinutes() + 1);

    // Search up to 1 year ahead
    const maxIterations = 366 * 24 * 60;

    for (let i = 0; i < maxIterations; i++) {
      if (this.matches(expression, current)) {
        return current;
      }
      current.setMinutes(current.getMinutes() + 1);
    }

    return null;
  }
}

// ============================================================================
// Scheduler Implementation
// ============================================================================

export class Scheduler {
  private jobs: Map<string, ScheduledJob> = new Map();
  private running = false;
  private intervalId: ReturnType<typeof setInterval> | null = null;
  private callbacks: JobCallback[] = [];

  constructor(private checkIntervalMs: number = 60000) {}

  /**
   * Add a scheduled job.
   */
  addJob(job: ScheduledJob): void {
    // Calculate next run time
    job.nextRun = CronParser.nextRun(job.schedule);
    this.jobs.set(job.id, job);
  }

  /**
   * Remove a scheduled job.
   */
  removeJob(jobId: string): boolean {
    return this.jobs.delete(jobId);
  }

  /**
   * Get a scheduled job by ID.
   */
  getJob(jobId: string): ScheduledJob | undefined {
    return this.jobs.get(jobId);
  }

  /**
   * List all scheduled jobs.
   */
  listJobs(): ScheduledJob[] {
    return Array.from(this.jobs.values());
  }

  /**
   * Register a callback for when a job is due.
   */
  onJobDue(callback: JobCallback): void {
    this.callbacks.push(callback);
  }

  /**
   * Start the scheduler.
   */
  start(): void {
    if (this.running) return;

    this.running = true;

    // Calculate time until next minute
    const now = new Date();
    const msUntilNextMinute = (60 - now.getSeconds()) * 1000 - now.getMilliseconds();

    // Start checking at the next minute boundary
    setTimeout(() => {
      this.checkJobs();
      this.intervalId = setInterval(() => this.checkJobs(), this.checkIntervalMs);
    }, msUntilNextMinute);
  }

  /**
   * Stop the scheduler.
   */
  stop(): void {
    this.running = false;
    if (this.intervalId) {
      clearInterval(this.intervalId);
      this.intervalId = null;
    }
  }

  /**
   * Check and run due jobs.
   */
  private async checkJobs(): Promise<void> {
    const now = new Date();

    for (const job of this.jobs.values()) {
      if (!job.enabled) continue;
      if (!job.nextRun || now < job.nextRun) continue;

      // Job is due - execute callbacks
      for (const callback of this.callbacks) {
        try {
          await callback(job);
        } catch (error) {
          console.error(`Error executing job ${job.id}:`, error);
        }
      }

      // Update job state
      job.lastRun = now;
      job.runCount++;
      job.nextRun = CronParser.nextRun(job.schedule, now);
    }
  }

  /**
   * Run due jobs once (non-blocking check).
   */
  async runOnce(): Promise<Map<string, Date>> {
    const results = new Map<string, Date>();
    const now = new Date();

    for (const job of this.jobs.values()) {
      if (!job.enabled) continue;
      if (!job.nextRun || now < job.nextRun) continue;

      // Job is due
      for (const callback of this.callbacks) {
        try {
          await callback(job);
        } catch (error) {
          console.error(`Error executing job ${job.id}:`, error);
        }
      }

      job.lastRun = now;
      job.runCount++;
      job.nextRun = CronParser.nextRun(job.schedule, now);
      results.set(job.id, now);
    }

    return results;
  }

  /**
   * Check if scheduler is running.
   */
  isRunning(): boolean {
    return this.running;
  }
}

// ============================================================================
// Helper Functions
// ============================================================================

export function createJob(
  id: string,
  workflowPath: string,
  schedule: string,
  inputs: Record<string, unknown> = {}
): ScheduledJob {
  return {
    id,
    workflowPath,
    schedule,
    timezone: 'UTC',
    enabled: true,
    lastRun: null,
    nextRun: CronParser.nextRun(schedule),
    runCount: 0,
    inputs,
  };
}
